keps/sig-network/2365-ingressclass-namespaced-params/README.md

```diff
@@ -0,0 +1,255 @@
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This KEP proposes adding an optional Namespace field to the IngressClass
+ParametersRef field.
+
+## Motivation
+
+After the initial release of IngressClass, a number of use cases called for the
+ability to reference namespace-scoped Parameters. For example, one could use a
+GatewayClass parameters CR to describe how and where a controller should be
+provisioned. This same thought process was also happening in the Service APIs
+subproject. It was ultimately deemed worthwhile for GatewayClass if we could
+also gain approval for a parallel API change to IngressClass.
+
+### Goals
+
+- Allow referencing namespace-scoped Parameters resources.
+
+### Non-Goals
+
+- Requiring all Parameters resources to be namespace-scoped.
+
+## Proposal
+
+Add an optional Namespace field to the IngressClass ParametersRef field.
+
+### Risks and Mitigations
+
+The option to reference namespace-scoped Parameters resources could lead to
+confusion. It is relatively rare for resource references to be able to target
+both cluster-scoped and namespace-scoped resources. We believe that the
+advantages of this KEP outweigh this potential confusion.
+
+## Design Details
+
+This will result in adding a new `IngressClassParametersReference` type that
+closely mirrors the existing `TypedLocalObjectReference` type that is currently
+in use.
+
+```golang
+// IngressClassParametersReference identifies an API object. This can be used
+// to specify a cluster-scoped or namespace-scoped resource.
+type IngressClassParametersReference struct {
+  // APIGroup is the group for the resource being referenced. If APIGroup is not
+  // specified, the specified Kind must be in the core API group. For any other
+  // third-party types, APIGroup is required.
+	// +optional
+	APIGroup *string
+	// Kind is the type of resource being referenced.
+	Kind string
+	// Name is the name of resource being referenced.
+  Name string
+  // Namespace is the namespace of the resource being referenced. When empty,
+  // this reference will refer to a cluster-scoped resource.
+  // +optional
+  Namespace string
+}
+```
+
+Use of this new `Namespace` field will be guarded a new
+`IngressClassNamespacedParams` feature gate.
```

Comment:
Question: How is this going to be incorporated into [IngressClassSpec](https://github.com/kubernetes/kubernetes/blob/5cd694ba7d7cb04614837727172fa297d8780edb/pkg/apis/networking/types.go#L291-L306) struct in a non-breaking fashion?

```go
	// Parameters is a link to a custom resource containing additional
	// configuration for the controller. This is optional if the controller does
	// not require extra parameters.
	// +optional
	Parameters *api.TypedLocalObjectReference
```

Reply:
That's a great question. My understanding was that changes to underlying Go types were acceptable between releases. I could be wrong on that though. Maybe @thockin knows if changing `Parameters *api.TypedLocalObjectReference` to `Parameters *IngressClassNamespacedParams` would be a breaking change here.

Reply:
The Go type is not really part of the API (though it does get exposed in client-go).  I think this is why folks like @deads2k are giving guidance to COPY the type you want into your own scope.

I don't know the client-go rules on this sort of 'breaking" change, but ISTR it was somewhat looser than the k8s APIs.

@lavalamp ?

Reply:
@DirectXMan12 - this might be something an IDL could automate - "clone a library type into my group" ?

Reply:
> The Go type is not really part of the API (though it does get exposed in client-go). I think this is why folks like @deads2k are giving guidance to COPY the type you want into your own scope.

TIL. I had wrongly assumed it to be part of the API.
I do recall to have ran into issues where breaking changes in client-go required some minor refactors in controllers using it. Although in my limited experience, the changes I've seen up to this point have been in client-go itself and not k8s API types.

Reply:
The thing we make promises about is the wire format of the API. So the question is what will old/new clients do with the new/old api? That probably works out OK (although I didn't review this super carefully).

Do we care about breaking the public go interface? kinda sorta. We often find a "good reason" to break it. :/ We don't typically consider that when looking at API changes. This would be less problematic if we ever managed to factor the client to split the api version into a separate package from all the logic.